src/main/java/io/vertx/core/parsetools/impl/LengthFieldParserImpl.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2011-2021 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.parsetools.impl;
+
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.impl.Arguments;
+import io.vertx.core.parsetools.LengthFieldParser;
+import io.vertx.core.parsetools.RecordParser;
+import io.vertx.core.streams.ReadStream;
+
+/**
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+
+public class LengthFieldParserImpl implements LengthFieldParser {
+
+  private Handler<Throwable> exceptionHandler;
+  private final RecordParser parser;
+
+  private long frameLength = -1;
+  private final int length;
+  private final int offset;
+  private final int max;
+
+  public LengthFieldParserImpl(int length, int offset, int max, ReadStream<Buffer> stream) {
+    Arguments.require(length == 1 || length == 2 || length == 3 || length == 4 || length == 8,
+      "Field length must be 1, 2, 3, 4, or 8");
+    Arguments.require(offset >= 0, "Field offset must be >= 0");
+    Arguments.require(max > 0, "Max frame length must be > 0");
+    this.length = length;
+    this.offset = offset;
+    this.max = max;
+    this.parser = RecordParser.newFixed(length + offset, stream);
+  }
+
+  @Override
+  public LengthFieldParser exceptionHandler(Handler<Throwable> handler) {
+    if (handler != null) {
+      exceptionHandler = handler;
+    }
+    return this;
+  }
+
+  @Override
+  public LengthFieldParser handler(Handler<Buffer> handler) {
+    if (handler == null) {
+      parser.handler(null);
+      parser.exceptionHandler(null);
+      parser.endHandler(null);
+      return this;
+    }
+    parser.handler(buffer -> {
+      try {
+        if (frameLength == -1) {
+          frameLength = handleFrameLength(buffer);
+          if (frameLength > max || frameLength <= 0) {
+            try {
+              String err = frameLength <= 0 ? "Frame length : " + frameLength + " <= 0" :
+                "Frame length is too large current: " + frameLength + " max: " + max;
+              IllegalStateException ex = new IllegalStateException(err);
+              if (exceptionHandler != null) {
+                exceptionHandler.handle(ex);
+              } else {
+                throw ex;
+              }
+            } finally {
+              frameLength = 0;
+            }
+          } else {
+            parser.fixedSizeMode((int) frameLength);
+            fetch(1);
```

Comment:
Are you sure this is needed?

Reply:
This is the weird part which confuses everyone. If I did not add it the `LengthFieldParser#fetch()` in general won't behave as expected!

Reply:
Another thing that I should clarify is that when the user invokes `fetch(1)` the use won't get the first item in the queue even if there are more than one item so it's important to have it in the parser to keep the chronological order of the items that needs to be fetched once it's there.